File: 977-distinct-subsequences-ii/distinct-subsequences-ii.java
class Solution {
    int MOD = 1_000_000_007;
    public int distinctSubseqII(String s) {
        int[] dp=new int[s.length()];
        Arrays.fill(dp,-1);
        return (int)subs(s,dp,0)%MOD;
    }
    public int subs(String s,int [] dp,int i){
        if(i>=s.length()){
            return 0;
        }
        if(dp[i]!=-1){
            return dp[i];
        }
        int ans=0;
        HashSet<Character> st=new HashSet<>();
        for(int idx=i;idx<s.length();idx++){
            if(!st.contains(s.charAt(idx))){
                st.add(s.charAt(idx));
                ans=(ans%MOD+1+subs(s,dp,idx+1))%MOD;
            }
        }
        return dp[i]=ans;
    }
}

File: 977-distinct-subsequences-ii/distinct-subsequences-ii.cpp
class Solution {
public:
    int mod=1000000007;
    int distinctSubseqII(string s) {
        vector<long long> dp(s.size(),-1);
        return count(s,0,dp);
    }
    long long count(string &s,int idx,vector<long long> &dp){
        if(s.size()<=idx){
            return 0;
        }
        if(dp[idx]!=-1){
            return dp[idx];
        }
        long long ans=0;
        unordered_set<int> st;
        for(int i=idx;i<s.size();i++){
            if(!st.count(s[i])){
                st.insert(s[i]);
                ans=(ans+1+count(s,i+1,dp))%mod;
                // st.erase(s[i]);
            }
        }
        return dp[idx]=ans;
    }
};